In [1]:
# Install packages (run once in your notebook environment)
!pip install pandas numpy scikit-learn scipy matplotlib wordcloud transformers torch nltk

In [2]:
# ============================================================
# 1) Imports & downloads
# ============================================================
import re, json
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from wordcloud import WordCloud

from transformers import pipeline
from scipy import sparse
from sklearn.base import BaseEstimator, TransformerMixin
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report, confusion_matrix
from sklearn.ensemble import RandomForestClassifier 
from sklearn.base import BaseEstimator, TransformerMixin

import nltk
nltk.download("stopwords", quiet=True)
nltk.download("wordnet", quiet=True)


C:\Users\vedah\anaconda3\Lib\site-packages\transformers\utils\generic.py:260: FutureWarning: `torch.utils._pytree._register_pytree_node` is deprecated. Please use `torch.utils._pytree.register_pytree_node` instead.
  torch.utils._pytree._register_pytree_node(


True

In [3]:
# ============================================================
# 2) Load your dataset (follow your reference)
# ============================================================
df = pd.read_csv("cleaned_dataset.csv")  

# Drop junk unnamed columns
df = df.drop(columns=[c for c in df.columns if str(c).startswith("Unnamed")], errors="ignore")
df["text"] = df["text"].fillna("")

print("=== Data overview ===")
print(df.info())
print(df.head(3))

=== Data overview ===
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 312 entries, 0 to 311
Data columns (total 6 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   business_name    312 non-null    object
 1   text             312 non-null    object
 2   rating           312 non-null    int64 
 3   rating_category  312 non-null    object
 4   label            312 non-null    object
 5   processed_text   312 non-null    object
dtypes: int64(1), object(5)
memory usage: 14.8+ KB
None
    business_name                                               text  rating  \
0  Abidin Tantuni  It was too oily and the place had problems wit...       2   
1  Abidin Tantuni  It was too spicy on my first try; but after a ...       4   
2  Abidin Tantuni  it tasted so good that I even ordered and ate ...       5   

      rating_category         label  \
0               taste  Clean Review   
1  outdoor_atmosphere  Clean Review   
2   indoor_atmospher

In [4]:
# ============================================================
# 3) Zero-shot + rule overrides
# ============================================================
labels = ["Advertisement", "Irrelevant content", "Rant without visit", "Clean review"]

clf_zs = pipeline("zero-shot-classification",
                  model="typeform/distilbert-base-uncased-mnli",
                  device=-1)  # CPU

def zero_shot_batch(texts, batch_size=16):
    out = []
    for i in range(0, len(texts), batch_size):
        chunk = texts[i:i+batch_size]
        res = clf_zs(chunk, candidate_labels=labels,
                     hypothesis_template="This review is {}.")
        if isinstance(res, dict): res = [res]
        out.extend([r["labels"][0] for r in res])
    return out

df["policy_pred_zeroshot"] = zero_shot_batch(df["text"].tolist(), batch_size=16)

# --- Rule-based overrides ---
PROMO_WORDS = r"(discount|coupon|promo|promotion|deal|sale|use code|visit|voucher|free)"
ANTI_VISIT = r"(never been|haven'?t been|did(n't| not) visit|did(n't| not) go|not been there)"
URL_PAT = r"(http[s]?://|www\.)"

def override_policy(raw_text, current_label):
    s = str(raw_text)
    if re.search(URL_PAT, s, re.I) or re.search(PROMO_WORDS, s, re.I):
        return "Advertisement"
    if re.search(ANTI_VISIT, s, re.I):
        return "Rant without visit"
    return current_label

df["policy_final"] = [override_policy(t, p) for t, p in zip(df["text"], df["policy_pred_zeroshot"])]
print(df["policy_final"].value_counts())


The `xla_device` argument has been deprecated in v4.4.0 of Transformers. It is ignored and you can safely remove it from your `config.json` file.
The `xla_device` argument has been deprecated in v4.4.0 of Transformers. It is ignored and you can safely remove it from your `config.json` file.
The `xla_device` argument has been deprecated in v4.4.0 of Transformers. It is ignored and you can safely remove it from your `config.json` file.
The `xla_device` argument has been deprecated in v4.4.0 of Transformers. It is ignored and you can safely remove it from your `config.json` file.
The `xla_device` argument has been deprecated in v4.4.0 of Transformers. It is ignored and you can safely remove it from your `config.json` file.


policy_final
Clean review          123
Rant without visit     91
Advertisement          87
Irrelevant content     11
Name: count, dtype: int64


In [5]:
# ============================================================
# 4) Choose training target
# ============================================================
TARGET_COL = "label" if "label" in df.columns and df["label"].notna().any() else "policy_final"
print(f"Training target = {TARGET_COL}")
y = df[TARGET_COL].astype(str)


Training target = label


In [6]:
# ============================================================
# 5) Split data 80-20
# ============================================================
X_train, X_test, y_train, y_test = train_test_split(
    df["text"], y, test_size=0.2, random_state=20, stratify=y
)


In [7]:
# ============================================================
# 6) Features (TF-IDF + simple regex)
# ============================================================
PHONE_PAT = r"\b\+?\d[\d\s\-]{6,}\b"
VISIT_WORDS = r"(i visited|we went|queued|ordered|ate|table|bill|waiter|menu)"

def simple_feats(text):
    t = str(text).lower()
    return {
        "has_url": int(bool(re.search(URL_PAT, t))),
        "has_phone": int(bool(re.search(PHONE_PAT, t))),
        "promo_hit": int(bool(re.search(PROMO_WORDS, t))),
        "exclam": str(text).count("!"),
        "allcaps_ratio": (sum(ch.isupper() for ch in str(text)) / max(1, len(str(text)))),
        "anti_visit": int(bool(re.search(ANTI_VISIT, t))),
        "visit_words": int(bool(re.search(VISIT_WORDS, t))),
        "relevance_hint": 1 - int(bool(re.search(PROMO_WORDS, t))),
    }

class FeatExtractor(BaseEstimator, TransformerMixin):
    KEYS = [
        "has_url","has_phone","promo_hit","exclam",
        "allcaps_ratio","anti_visit","visit_words","relevance_hint"
    ]
    def fit(self, X, y=None):
        return self
    def transform(self, X):
        # X is an iterable of strings (Series/list)
        rows = [simple_feats(x) for x in X]
        # Build matrix in a stable, explicit order of KEYS
        return np.array([[r[k] for k in self.KEYS] for r in rows], dtype=float)

# Text features
from sklearn.feature_extraction.text import TfidfVectorizer
tfidf = TfidfVectorizer(ngram_range=(1,2), max_features=30000, min_df=2)

X_train_tfidf = tfidf.fit_transform(X_train)
X_test_tfidf  = tfidf.transform(X_test)

# Numeric features
num_extractor = FeatExtractor()
X_train_num = num_extractor.fit_transform(X_train)
X_test_num  = num_extractor.transform(X_test)

# Combine (sparse hstack keeps memory low)
X_train_all = sparse.hstack([X_train_tfidf, sparse.csr_matrix(X_train_num)], format="csr")
X_test_all  = sparse.hstack([X_test_tfidf,  sparse.csr_matrix(X_test_num)],  format="csr")

print("Train shape:", X_train_all.shape, "| Test shape:", X_test_all.shape)


Train shape: (249, 922) | Test shape: (63, 922)


In [8]:
# ============================================================
# 7) Train classifier
# ============================================================
clf = LogisticRegression(max_iter=2000, class_weight="balanced")
clf.fit(X_train_all, y_train)

LogisticRegression(class_weight='balanced', max_iter=2000)

In [16]:
# ============================================================
# 8) Evaluate
# ============================================================
print("\n=== RICH MODEL: TF-IDF(3000) + Logistic Regression + numeric regex features===")
y_pred = clf.predict(X_test_all)
print(classification_report(y_test, y_pred, digits=3))


=== RICH MODEL: TF-IDF(3000) + Logistic Regression + numeric regex features===
                      precision    recall  f1-score   support

       Advertisement      0.889     1.000     0.941         8
        Clean Review      0.975     0.951     0.963        41
  Irrelevant Content      0.800     1.000     0.889         8
Review without Visit      1.000     0.667     0.800         6

            accuracy                          0.937        63
           macro avg      0.916     0.904     0.898        63
        weighted avg      0.944     0.937     0.935        63



Accuracy:93.7%

Macro Average F1: At 0.898, the macro F1 shows that performance is consistently high across all classes.
Weighted Average F1: The weighted F1 of 0.935 indicates the model maintains strong results even when considering class imbalance.

**Per Class Analysis**

*Advertisement*
- Precision = 0.889, Recall = 1.000, F1 = 0.941
The model correctly identifies all advertisements (perfect recall). A few non-ads are misclassified as ads, which slightly lowers precision, but overall detection is highly reliable.

*Clean Review*
- Precision = 0.975, Recall = 0.951, F1 = 0.963
Clean reviews are detected with both high precision and recall. This is the best-performing class, showing the model is excellent at recognizing genuine reviews.

*Irrelevant Content*
- Precision = 0.800, Recall = 1.000, F1 = 0.889
All irrelevant reviews are captured (perfect recall), though some clean reviews are mistakenly flagged as irrelevant. This trade-off lowers precision, but ensures that irrelevant content is not missed.

*Review without Visit*
- Precision = 1.000, Recall = 0.667, F1 = 0.800
The model is very cautious with this class, only labeling reviews as “No Visit” when it is highly confident (precision 100%). However, recall is weaker, with one-third of such reviews being missed.

In [14]:
TEXT_COL = "text"

In [11]:
# Build df2 to mirror your style, using the chosen target
df2 = df[[TEXT_COL, TARGET_COL]].rename(columns={TEXT_COL: "text", TARGET_COL: "label"}).copy()
df2["text"] = df2["text"].fillna("")
df2["label"] = df2["label"].astype(str)

# Vectorize text (baseline)
vectorizer_base = TfidfVectorizer(max_features=3000, stop_words="english")
X_base = vectorizer_base.fit_transform(df2["text"])
y_base = df2["label"]

# Split
X_train_b, X_test_b, y_train_b, y_test_b = train_test_split(
    X_base, y_base, test_size=0.2, random_state=20, stratify=y_base
)

# Logistic Regression baseline
logreg_base = LogisticRegression(
    max_iter=2000,
    class_weight="balanced",
    solver="lbfgs",
    multi_class="auto"
)
logreg_base.fit(X_train_b, y_train_b)

# Evaluate
y_pred_b = logreg_base.predict(X_test_b)
print("\n=== BASELINE: TF-IDF(3000) + Logistic Regression ===")
print(classification_report(y_test_b, y_pred_b, digits=3))


=== BASELINE: TF-IDF(3000) + Logistic Regression ===
                      precision    recall  f1-score   support

       Advertisement      0.889     1.000     0.941         8
        Clean Review      0.889     0.976     0.930        41
  Irrelevant Content      0.833     0.625     0.714         8
Review without Visit      1.000     0.500     0.667         6

            accuracy                          0.889        63
           macro avg      0.903     0.775     0.813        63
        weighted avg      0.892     0.889     0.879        63



**Logistic Regression**

Accuracy: 88.9%

- Macro average F1: 0.813 which shows that performance across classes is uneven; some classes do much better than others.

- Weighted average F1: 0.879 → shows performance is better on frequent classes (like Clean Review) but weaker on rare ones.


**Per class Analysis** 

*Advertisement* 

- Advertisement Precision = 0.889, Recall = 1.000, F1 = 0.941 
The model catches all ads (recall 100%) but a few non-ads are incorrectly predicted as ads (precision <1). Very strong performance here.

*Clean review*

- Precision = 0.889, Recall = 0.976, F1 = 0.930 
Almost all clean reviews are correctly identified. Small precision loss means a handful of non-clean reviews are mislabeled as clean. Overall, this is the best performing class.

*Irrelevant content* 
 
- Precision = 0.833, Recall = 0.625, F1 = 0.714
When the model says “irrelevant,” it’s usually right (precision good). But it misses ~40% of actual irrelevant reviews (low recall). This suggests that the model is not picking up enough signals for this class(maybe due to too few training examples or overlap with “Clean Review”.)

*Rant without visit*

- Precision = 1.000, Recall = 0.500, F1 = 0.667
The model is super cautious as it only labels a review as “No Visit” when it’s very sure (precision = 100%). But it misses half the actual “No Visit” cases (recall 0.5). This is typical when the class has low support as the model learns too little.

## Overall Conclusion

**Baseline Model** 

- Accuracy: 88.9%
- Macro F1: 0.813 → performance uneven, weaker on minority classes
- Weighted F1: 0.879 → strong overall, biased toward frequent classes (Clean Review)

**Rich Model**
- Accuracy: 93.7%
- Macro F1: 0.898 which is consistently high across all classes
- Weighted F1: 0.935 → better balance across frequent and rare categories

Summary: The rich model clearly improves accuracy and balances performance across classes compared to the baseline.